txircd/modules/extra/webirc.py

```diff
@@ -0,0 +1,57 @@
+from twisted.plugin import IPlugin
+from twisted.words.protocols import irc
+from txircd.config import ConfigValidationError
+from txircd.module_interface import ICommand, IModuleData, Command, ModuleData
+from txircd.utils import isValidHost
+from zope.interface import implements
+
+irc.ERR_ALREADYREGISTERED = "462"
+
+class WebIRC(ModuleData, Command):
+	implements(IPlugin, IModuleData, ICommand)
+
+	name = "WebIRC"
+
+	def userCommands(self):
+		return [ ("WEBIRC", 1, self) ]
+
+	def verifyConfig(self, config):
+		if "webirc_allowed_sources" in config:
+			if not isinstance(config["webirc_allowed_sources"], dict):
+				raise ConfigValidationError("webirc_allowed_sources", "value must be a dictionary")
+			for ip, password in config["webirc_allowed_sources"]:
+				if not isinstance(ip, basestring):
+					raise ConfigValidationError("webirc_allowed_sources", "ip value must be a string")
+				if not isinstance(password, basestring):
+					raise ConfigValidationError("webirc_allowed_sources", "password value must be a string")
+
+	def parseParams(self, user, params, prefix, tags):
+		if user.isRegistered():
+			user.sendSingleError("WebircCmd", irc.ERR_ALREADYREGISTERED, "WEBIRC may only be used during registration")
+			return None
+		if len(params) < 4:
+			user.sendSingleError("WebircCmd", irc.ERR_NEEDMOREPARAMS, "WEBIRC", "Not enough parameters")
+			return None
+		if user.ip not in self.ircd.config.get("webirc_allowed_sources", {}) and user.realHost not in self.ircd.config.get("webirc_allowed_sources", {}):
+			self.ircd.log.warn("WEBIRC was requested from IP \"{ip}\", but the IP does not match any WEBIRC configuration.", ip=user.ip)
+			return None
+		if self.ircd.config["webirc_allowed_sources"][user.ip] != params[0]:
+			self.ircd.log.warn("WEBIRC was requested from IP \"{ip}\" with password \"{password}\", but this password does not match the WEBIRC configuration for this IP.", ip=user.ip, password=params[0])
+			return None
+		# We don't need params[1]; this is the client name which we don't use. We verify that the DNS resolution is correct and set the provided IP as the host if it is incorrect.
+		host = params[2]
+		ip = params[3]
+		if not isValidHost(host) or len(host) > self.ircd.config.get("hostname_length", 64) or user.resolveDNS(params[3]) != host:
+			self.ircd.log.warn("DNS resolution for WEBIRC command from IP \"{ip}\" with requested IP \"{requestip}\" and requested host \"{requesthost}\" has failed; using the requested IP address as the host instead.", ip=user.ip, requestip=ip, requesthost=host)
```

Comment:
For logging, you should use `{user.ip}` in the string and `user=user` as a parameter. Same applies to all of the other log calls above and below.